File: 111.minimum-depth-of-binary-tree.py
#
# @lc app=leetcode id=111 lang=python3
#
# [111] Minimum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
from typing import *
class Solution: # pay attention to the situation of [1,2]
    def __init__(self):
        self.min_depth = 100000

    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        self.helper(root,1)
        return self.min_depth 

    def helper(self,root,depth):
        if not root.left and not root.right:
            if depth < self.min_depth:
                self.min_depth = depth
            return
        if root.left:
            self.helper(root.left,depth +1 )
        if root.right:
            self.helper(root.right,depth +1)
        
# @lc code=end

File: 111.minimum-depth-of-binary-tree.go
/*
 * @lc app=leetcode id=111 lang=golang
 *
 * [111] Minimum Depth of Binary Tree
 */
package main

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// 看下代码怎么组织
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	if root.Left == nil && root.Right == nil {
		return 1
	}
	if root.Left == nil {
		return minDepth(root.Right) + 1
	}
	if root.Right == nil {
		return minDepth(root.Left) + 1
	}
	leftDepth := minDepth(root.Left)
	rightDepth := minDepth(root.Right)
	// 这里是个边界条件
	if leftDepth < rightDepth {
		return leftDepth + 1
	}
	return rightDepth + 1
}

// @lc code=end
